google/resource_attached_disk.go

```diff
@@ -0,0 +1,286 @@
+package google
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	compute "google.golang.org/api/compute/v1"
+)
+
+func resourceAttachedDisk() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAttachedDiskCreate,
+		Read:   resourceAttachedDiskRead,
+		Update: resourceAttachedDiskUpdate, // TODO remove this method? make sure it's not called? etc
+		Delete: resourceAttachedDiskDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: resourceAttachedDiskImport,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(300 * time.Second),
+			Update: schema.DefaultTimeout(240 * time.Second),
+			Delete: schema.DefaultTimeout(240 * time.Second),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"attached_disk": {
+				Type:     schema.TypeString,
+				Required: true,
+				// TODO (chrisst) - add force new here so that people have to recreate instead of update
+				// TODO (chrisst) change the name of this attribute
+			},
+			"attached_instance": {
+				Type:     schema.TypeString,
+				Required: true,
+				// TODO (chrisst) - add force new here so that people have to recreate instead of update
+				// TODO (chrisst) change the name of this attribute
+			},
+			"project": {
+				// Adding project as computed so that import can use parseImportId
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"zone": {
+				// Adding zone as computed so that import can use parseImportId
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"type": {
+				// TODO play with this to determine if somebody can change between PERSISTANT vs SCRATCH.
+				// And see if it's worth allowing a user to pass in
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"self_link": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAttachedDiskCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	instance := d.Get("attached_instance").(string)
+	zone, err := getZoneForAttachedDisk(d, config, instance)
+	if err != nil {
+		return err
+	}
+	instanceName := GetResourceNameFromSelfLink(instance)
+	diskName := GetResourceNameFromSelfLink(d.Get("attached_disk").(string))
+
+	attachedDisk := compute.AttachedDisk{}
+	attachedDisk.DeviceName = diskName
+	// TODO strip the v1 part of the path from this instead of the full url
+	attachedDisk.Source = fmt.Sprintf("/projects/%s/zones/%s/disks/%s",
+		project, zone, diskName)
+
+	op, err := config.clientCompute.Instances.AttachDisk(project, zone, instanceName, &attachedDisk).Do()
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s", instanceName, diskName))
```

Comment:
If you also put the project and zone in the ID then you can save yourself a bunch of code later in the Read() and Delete() funcs, because then all you need to do to get the info you need would be to parse the ID.

Reply:
🤔 I'm intrigued by this. It would mean only allowing `project/zone/instance:disk` and the more verbose url'y version in the import but I think that's fine.